provider.go

```diff
@@ -0,0 +1,237 @@
+package saramakeycloak
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/Nerzal/gocloak"
+	"github.com/Shopify/sarama"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+var (
+	// ErrAlreadyClosed means that Provider was already closed before.
+	ErrAlreadyClosed = errors.New("provider is already closed")
+)
+
+type token struct {
+	jwt       *gocloak.JWT
+	expiresAt time.Time // absolute time when access token expires.
+}
+
+func newToken(jwt *gocloak.JWT) *token {
+	return &token{
+		jwt:       jwt,
+		expiresAt: time.Now().Add(time.Duration(jwt.ExpiresIn) * time.Second),
+	}
+}
+
+func (t *token) activeNow() bool {
+	if t == nil {
+		return false
+	}
+
+	return time.Now().Before(t.expiresAt)
+}
+
+func (t *token) toSaramaToken() *sarama.AccessToken {
+	return &sarama.AccessToken{Token: t.jwt.AccessToken}
+}
+
+// Provider is used to use keycloak as authentication provider for kafka.
+// It implements sarama.AccessTokenProvider interface.
+type Provider struct {
+	logger *zap.Logger
+
+	keycloak        Keycloak
+	keycloakTimeout time.Duration
+
+	// credentials to authenticate in keycloak.
+	clientID     string
+	clientSecret string
+	realm        string
+
+	refreshThreshold time.Duration
+
+	tokenMu   *sync.RWMutex // guards token.
+	tokenCond *sync.Cond
+	token     *token // current active token.
+
+	stopOnce sync.Once
+	stopCh   chan struct{}
+}
+
+// New returns new Provider.
+func New(c Config) (*Provider, error) {
+	c.applyDefaults()
+
+	if err := c.validate(); err != nil {
+		return nil, errors.Wrap(err, "config validation failed")
+	}
+
+	mu := &sync.RWMutex{}
+
+	p := &Provider{
+		refreshThreshold: c.RefreshThreshold,
+		keycloak:         &keycloakWithMetrics{gocloak.NewClient(c.KeycloakHostPort)},
+		keycloakTimeout:  c.KeycloakTimeout,
+		clientID:         c.ClientID,
+		clientSecret:     c.ClientSecret,
+		realm:            c.Realm,
+		logger:           c.Logger,
+		tokenMu:          mu,
+		tokenCond:        sync.NewCond(mu),
+		stopCh:           make(chan struct{}),
+		stopOnce:         sync.Once{},
+	}
+
+	go p.updateLoop()
+
+	return p, nil
+}
+
+// setClient sets internal keycloak client.
+// must be used only for testing!
+func (p *Provider) setClient(c Keycloak) {
+	p.keycloak = c
+}
+
+func (p *Provider) updateLoop() {
+	onTokenUpdate := func(t *time.Timer, token *token) {
+		p.token = token
+		p.tokenCond.Broadcast()
+
+		if p.token != nil {
+			expiresIn := time.Duration(token.jwt.ExpiresIn) * time.Second
+			t.Reset(expiresIn - p.refreshThreshold)
+		} else {
+			t.Reset(100 * time.Millisecond)
```

Comment:
This retry timeout looks like a magic number

Reply:
Really, it is. I'll extract to the configuration.